tensorflow/python/ops/image_grad.py

```diff
@@ -153,3 +153,211 @@ def _CropAndResizeGrad(op, grad):
       grad, op.inputs[0], op.inputs[1], op.inputs[2])
 
   return [grad0, grad1, None, None]
+
+
+def _CustomReciprocal(x, my_eps=1e-10):
+  """
+  Performs reciprocal with an eps added to the input. This is to avoid
+  inversion errors or divide by zeros or NaNs.
+  Inputs:
+    x -> input tensor to be reciprocat-ed
+    my_eps -> custom machine precision epsilon
+  Returns:
+    x_reciprocal -> reciprocal of x added with my_eps
+  """
+  return reciprocal(x + my_eps)
+
+
+@ops.RegisterGradient("RGBToHSV")
+def _RGBToHSVGrad(op, grad):
+  """The gradients for `rgb_to_hsv` operation.
+  This function is a piecewise continuous function as defined here:
+  https://en.wikipedia.org/wiki/HSL_and_HSV#From_RGB
+  We perform the multi variate derivative and compute all partial derivates
+  seperately before adding them in the end. Formulas are given before each
+  partial derivative calculation.
+
+  Args:
+    op: The `rgb_to_hsv` `Operation` that we are differentiating.
+    grad: Gradient with respect to the output of the `rgb_to_hsv` op.
+
+  Returns:
+    Gradients with respect to the input of `rgb_to_hsv`.
+  """
+  # Input Channels
+  reds = op.inputs[0][..., 0]
+  greens = op.inputs[0][..., 1]
+  blues = op.inputs[0][..., 2]
+  # Output Channels
+  hue = op.outputs[0][..., 0]
+  saturation = op.outputs[0][..., 1]
+  value = op.outputs[0][..., 2]
+
+  # Mask/Indicator for max and min values of each pixel.
+  # Arbitrary assignment in case of tie breakers with R>G>B.
+  # Max values
+  red_biggest = cast((reds >= blues) & (reds >= greens), dtypes.float32)
+  green_biggest = cast((greens > reds) & (greens >= blues), dtypes.float32)
+  blue_biggest = cast((blues > reds) & (blues > greens), dtypes.float32)
+  # Min values
+  red_smallest = cast((reds < blues) & (reds < greens), dtypes.float32)
+  green_smallest = cast((greens <= reds) & (greens < blues), dtypes.float32)
+  blue_smallest = cast((blues <= reds) & (blues <= greens), dtypes.float32)
+
+  ##############################################################
+  #  Derivatives of R, G, B wrt Value slice
+  ##############################################################
+  dv_dr = red_biggest
+  dv_dg = green_biggest
+  dv_db = blue_biggest
+  ##############################################################
+  # Derivatives of R, G, B wrt Saturation slice
+  ##############################################################
+  # The first term in the addition is the case when the corresponding color
+  # from (r,g,b) was "MAX"
+  # -> derivative = MIN/square(MAX), MIN could be one of the other two colors
+  # The second term is the case when the corresponding color from
+  # (r,g,b) was "MIN"
+  # -> derivative = -1/MAX, MAX could be one of the other two colours.
+  ds_dr = cast(reds > 0, dtypes.float32) * add(red_biggest * \
```

Comment:
Could we use a helper function to compute `ds_dr`, `ds_dr_1`, ... `ds_dr_5` to avoid the repeat code?

Reply:
I thought of doing that but the formula is different each time. Each of the 5 terms has different formula for each of the three colors. 
https://en.wikipedia.org/wiki/HSL_and_HSV#From_RGB

Reply:
Ah, I did not notice that.